<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Spats - Getting Started</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"></meta>
	<meta name="description" content="A brief guide to using Spats for RNA structure determination with SHAPE-Seq"></meta>
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen"></link>
  </head>
  <body>
    <div id="wrap">
     	<div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Spats</h1>
                  </a>
                  <h2>Read Processing and Reactivity Generation for SHAPE-Seq</h2>
                </td>
                <td align="right" valign="middle"> 
                    <a href="http://bio.math.berkeley.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/UCBerkeley-seal.scaled.gif" border="0">
                    </a>&nbsp;
                    <a href="http://http://www.mccormick.northwestern.edu/chemical-biological/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/Northwestern Formal_horizontal_shrunk.png" border="0">
                    </a>&nbsp;
                </td>
              </tr>
            </tbody>
          </table>
        </div>

      <div id="main">
       	<div id="subheader">
	        <table width="100%">
	          <tbody>
	            <tr>
	              <td> <strong>Spats</strong> processes reads and calculates SHAPE reactivities for SHAPE-Seq
	              experiments on multiple RNAs. It accepts raw paired-end sequencing reads in FASTQ format, and a target
	              sequence file containing the sequences of RNAs present in the experimental pool. Spats then performs
	              read alignment to calculate distributions of read ends in the SHAPE (+) and (-) channel for each
	              nucleotide in each RNA. Spats then estimates nucleotide resolution SHAPE reactivities for each RNA,
	              using a model-driven maximum likelihood procedure based on a model of the reverse transcriptase
	              process used in the SHAPE-Seq experiment.	
	                Spats is a collaborative effort between the
					 <a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a> at UC Davis, the <a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a> at UC Berkeley, the 
	                <a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a>
	                 at the University of Washington, and
	                <a href="http://luckslab.org/">the Lucks Lab</a>
	                at Northwestern University. <br>
	                <br>
	                Spats is provided under the OSI-approved <a
	                  href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost
	                  License</a>. fastx_spats_toolkit_v_0.1.0 is provided under the <a
	                  href="http://en.wikipedia.org/wiki/Affero_General_Public_License">AGPLv3
	                  License</a>. </td> 
	              <td align="right" valign="middle"> <a
	                  href="http://opensource.org"><img alt="Open Source
	                    Software" src="images/osi-certified.gif" border="0"></a>
	              </td>
	            </tr>
	          </tbody>
	        </table>
	      </div>
	
        <div id="rightside">
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="installation.html">Installation</a></li>
              <li><a href="manual.html">Manual</a></li>
            </ul>
          </div>

		            <h2><u>Releases</u></h2>
		          <div class="box">
		            <ul>
		              <table width="100%">
		                <tbody>
                                  <tr>
                                    <td>spats version 1.0.2</td>
                                    <td align="right">5/30/2017</td>
                                  </tr>
                                  <tr>
                                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v1.0.2">&nbsp;&nbsp;&nbsp;Source code</a></td>
                                  </tr>
                                  <tr>
                                    <td>spats version 1.0.1</td>
                                    <td align="right">9/10/2016</td>
                                  </tr>
                                  <tr>
                                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v1.0.1">&nbsp;&nbsp;&nbsp;Source code</a></td>
                                  </tr>
			          <tr>
		                    <td>spats version 1.0.0</td>
		                    <td align="right">3/21/2016</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v1.0.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>spats version 0.8.0</td>
		                    <td align="right">9/4/2014</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>fastx_spats_toolkit version 0.1.0</td>
		                    <td align="right">9/4/2014</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
						  <td>fastx_spats_toolkit version 0.1.0 - binary distribution for 64-bit GNU/Linux 2.6.18</td>
		                  <td align="right">9/4/2014</td>
						  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>spats version 0.7.0</td>
		                    <td align="right">4/10/2012</td>
		                  </tr>
		                  <tr>
		                    <td><a href="./downloads/spats-0.7.0.tar.gz"
		                        onclick="javascript:
		                        pageTracker._trackPageview('/downloads/cufflinks_source');
		                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                </tbody>
		              </table>
		            </ul>
		          </div>

          <h2>Publications</h2>
          <div class="box">
            <ul>

 			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Trapnell C, Lucks JB, Mortimer SA, Luo S, Schroth GP, Doudna JA, Arkin AP, Pachter L.<b> <a href="http://www.pnas.org/content/108/27/11069.long">Modeling and automation of sequencing-based characterization of RNA structure</b></a> <br/><i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106541108</p><br/></li>

		      <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Lucks JB, Pachter L.<b> <a href="http://arxiv.org/abs/1106.5061v2">RNA structure characterization from chemical mapping experiments</a></b> <br><i><a href="Aviran_Allerton_2011.pdf"></a>Proceedings of the 49th Allerton Conference on Communication, Control, and Computing, 1743-1750, 2011</i></p><br/></li>

			  <li style="font-size: x-small; line-height: 130%"><p>Lucks JB, Mortimer SA, Trapnell C, Luo S, Aviran S, Schroth GP, Pachter L, Doudna JA, Arkin AP. <b> <a href="http://www.pnas.org/content/108/27/11063.long">Multiplexed RNA structure characterization with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
				<i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106501108</p><br/>

			  <li style="font-size: x-small; line-height: 130%"><p>Mortimer SA, Trapnell C, Aviran S, Pachter L and Lucks JB. <b> <a href="http://onlinelibrary.wiley.com/doi/10.1002/9780470559277.ch120019/abstract">SHAPE-Seq High Throughput RNA Structure Analysis</a></b> <br/>
			  <i><a href="http://onlinelibrary.wiley.com/book/10.1002/9780470559277">Current Protocols in Chemical Biology</a></i>10.1002/9780470559277.ch120019</p><br/>

			  <li style="font-size: x-small; line-height: 130%"><p>Loughrey D, Watters KE, Settle AH, and Lucks JB. <b> <a href="http://dx.doi.org/10.1093/nar/gku909">SHAPE-Seq 2.0: Systematic Optimization and Extension of High-Throughput Chemical Probing of RNA Structure with Next Generation Sequencing</a></b> <br/>
			  <i><a href="http://dx.doi.org/10.1093/nar/gku909"></a>Nucleic Acids Research</i> 10.1093/nar/gku909</p><br/>
			  </li>
			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S and Pachter L. <b> <a href="http://rnajournal.cshlp.org/content/early/2014/10/20/rna.043844.113.short ">Rational experiment design for sequencing-based RNA structure mapping</a></b> <br/>
			  <i><a href="http://rnajournal.cshlp.org/content/early/2014/10/20/rna.043844.113.short ">RNA</a> 10.1261/rna.043844.113</i></p><br/>
			  </li>
			  <li style="font-size: x-small; line-height: 130%"><p>Watters KE, Abbott TR, and Lucks JB. <b> <a href="http://dx.doi.org/10.1093/nar/gkv879">Simultaneous characterization of cellular RNA structure and function with in-cell SHAPE-Seq</a></b> <br/>
			  <i><a href="http://dx.doi.org/10.1093/nar/gkv879"></a>Nucleic Acids Research</i> 10.1093/nar/gkv879</p><br/>
			  </li>
                          <li style="font-size: x-small; line-height: 130%"><p>Watters KE, Yu AM, Strobel EJ, Settle AH, Lucks JB.<b> <a href="http://dx.doi.org/10.1016/j.ymeth.2016.04.002">Characterizing RNA structures in vitro and in vivo with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
                          <i><a href="http://dx.doi.org/10.1016/j.ymeth.2016.04.002"></a>Methods</i> 10.1016/j.ymeth.2016.04.002 </p><br/>
                          </li>
                          <li style="font-size: x-small; line-height: 130%"><p>Watters KE*, Strobel EJ*, Yu AM, Lis JT, Lucks JB.<b> <a href="http://dx.doi.org/10.1038/nsmb.3316">Cotranscriptional folding of a riboswitch at nucleotide resolution</a></b> <br/>                          <i><a href="http://dx.doi.org/10.1038/nsmb.3316"></a>Nature Structure and Molecular Biology</i> 10.1038/nsmb.3316 </p><br/>
                          </li>
            </ul>
          </div>
          <h2>Other Contributors</h2>
          <div class="box">
            <ul>
              <li><a href="http://rna.berkeley.edu/">Stefanie Mortimer</a></li>
              <li>Shujun Luo</li>
              <li>Gary Schroth</li>
              <li>Adam Arkin</li>
		  	  <li><a href="http://rna.berkeley.edu/">Jennifer Doudna</a></li>
            </ul>
          </div>
          <h2>Links</h2>
          <div class="box">
            <ul>
			  <li><a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a></li>
			  <li><a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a></li>
              <li><a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a></li>
              <li><a href="http://luckslab.org/">Lucks Lab</a></li>
            </ul>
          </div>
        </div>

        <!-- End of "rightside" -->
        <div id="leftside">
          <table>
            <tbody>
              <tr>
                <td><h3>Installation</h3>
			  	  <br/>
					Spats runs on intel-based computers running Linux or Mac OS X. Currently spats can be run from the released binaries and can be built from the source code.<p></p></br>
<ul>
  <li><b>Requirements</b></li>
  <ul>
    <li>Bowtie 0.12.8, must be available in your <tt>$PATH</tt>. Get it at: <a href="http://sourceforge.net/projects/bowtie-bio/files/bowtie/0.12.8/">http://sourceforge.net/projects/bowtie-bio/files/bowtie/0.12.8/</a></li>
    <li><tt>cutadapt</tt>: python package (<tt>pip install cutadapt</tt>)</li>
  </ul>
  <li><b>Binary Distributions</b></li>
  <ul>
    <li>Available for Linux/x64 and Mac at: https://github.com/LucksLab/spats/releases/</li>
  </ul>
  <li><b>Build from Source</b></li>
  <ul>
    <li><tt>make -s</tt></li>
    <li>Add the <tt>bin</tt> directory to your <tt>$PATH</tt>.</li>
  </ul>
</ul>
<br/><br/><br/><br/><br/>                  
                  <div id="footer">
                    <table cellspacing="15" width="100%">
			          <tbody>
			            <tr>
			              <td> This research was supported in part by
			                the Miller Institute for Basic Research in Science at UC
			                Berkeley, the Alfred P. Sloan Foundation, and the National Institutes of Health. </td>
			              <td align="right"> Administrators: <a
			                  href="mailto:cole@cs.umd.edu">Cole Trapnell</a> and <a
				                  href="">Julius B. Lucks (j b lucks at northwestern dot edu)</a>.
			                Design by <a href="http://www.free-css-templates.com"
			                  title="Design by David Herreman">David Herreman</a> </td>
			            </tr>
			          </tbody>
			        </table>
                  </div>

                  <!-- Google analytics code -->
                  <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
                  <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
                  <!-- End google analytics code --> </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>
</html>
