<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Spats</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  	<div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Spats</h1>
                  </a>
                  <h2>Read Processing and Reactivity Generation for SHAPE-Seq</h2>
                </td>
                <td align="right" valign="middle"> 
                    <a href="http://bio.math.berkeley.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/UCBerkeley-seal.scaled.gif" border="0">
                    </a>&nbsp;
                    <a href="http://www.cheme.cornell.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/Cornell_logo.jpg" border="0">
                    </a>&nbsp;
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
  	<div id="subheader">
        <table width="100%">
          <tbody>
            <tr>
              <td> <strong>Spats</strong> processes reads and calculates SHAPE reactivities for SHAPE-Seq
              experiments on multiple RNAs. It accepts raw paired-end sequencing reads in fastq format, and a target
              sequence file containing the sequences of RNAs present in the experimental pool. Spats then performs
              read alignment to calculate distributions of read ends in the SHAPE (+) and (-) channel for each
              nucleotide in each RNA. Spats then estimates nucleotide resolution SHAPE reactivities for each RNA,
              using a model-driven maximum likelihood procedure based on a model of the reverse transcriptase
              process used in the SHAPE-Seq experiment.	
                Spats is a collaborative effort between the
				 <a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a> at UC Davis, the <a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a> at UC Berkeley, the 
                <a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a>
                 at the University of Washington, and
                <a href="http://luckslab.org/">the Lucks Lab</a>
                at Cornell University. <br>
                <br>
                Spats is provided under the OSI-approved <a
                  href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost
                  License</a>. fastx_spats_toolkit_v_0.1.0 is provided under the <a
                  href="http://en.wikipedia.org/wiki/Affero_General_Public_License">AGPLv3
                  License</a>. </td> 
              <td align="right" valign="middle"> <a
                  href="http://opensource.org"><img alt="Open Source
                    Software" src="images/osi-certified.gif" border="0"></a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
  <div id="main">
	
      <div id="rightside">
        <h2>Site Map</h2>
        <div class="box">
          <ul>
            <li><a href="index.html">Home</a></li>
            <li><a href="installation.html">Installation</a></li>
            <li><a href="manual.html">Manual</a></li>
			  <!-- <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li> -->
          </ul>
        </div>
        <!-- <h2>News and updates</h2>
                  <div class="box">
                    <ul>
                      <table width="100%">
                        <tbody>
                          <tr>
                            <td>New releases and related tools will be announced
                              through the <a
                                href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing


                                  list</b></a></td>
                          </tr>
                        </tbody>
                      </table>
                    </ul>
                  </div> -->

		  <!-- <h2>Getting Help</h2>
		            <div class="box">
		              <ul>
		                <table width="100%">
		                  <tbody>
		                    <tr>
		                      <td>Questions about Cufflinks should be sent to <a
		                          href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
		                        Please do not email technical questions to
		                        Cufflinks contributors directly.</td>
		                    </tr>
		                  </tbody>
		                </table>
		              </ul>
		            </div> -->

		          <a href="./downloads">
		            <h2><u>Releases</u></h2>
		          </a>
		          <div class="box">
		            <ul>
		              <table width="100%">
		                <tbody>
			              <tr>
		                    <td>spats version 1.0.0</td>
		                    <td align="right">3/21/2016</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>spats version 0.8.0</td>
		                    <td align="right">9/4/2014</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>fastx_spats_toolkit version 0.1.0</td>
		                    <td align="right">9/4/2014</td>
		                  </tr>
		                  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
						  <td>fastx_spats_toolkit version 0.1.0 - binary distribution for 64-bit GNU/Linux 2.6.18</td>
		                  <td align="right">9/4/2014</td>
						  <tr>
		                    <td><a href="https://github.com/LucksLab/spats/releases/tag/v0.8.0">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                  <tr>
		                    <td>spats version 0.7.0</td>
		                    <td align="right">4/10/2012</td>
		                  </tr>
		                  <tr>
		                    <td><a href="./downloads/spats-0.7.0.tar.gz"
		                        onclick="javascript:
		                        pageTracker._trackPageview('/downloads/cufflinks_source');
		                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
		                  </tr>
		                </tbody>
		              </table>
		            </ul>
		          </div>

        <h2>Publications</h2>
        <div class="box">
            <ul>

 			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Trapnell C, Lucks JB, Mortimer SA, Luo S, Schroth GP, Doudna JA, Arkin AP, Pachter L.<b> <a href="http://www.pnas.org/content/108/27/11069.long">Modeling and automation of sequencing-based characterization of RNA structure</b></a> <br/><i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106541108</p><br/></li>

		      <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Lucks JB, Pachter L.<b> <a href="http://arxiv.org/abs/1106.5061v2">RNA structure characterization from chemical mapping experiments</a></b> <br><i><a href="Aviran_Allerton_2011.pdf"></a>Proceedings of the 49th Allerton Conference on Communication, Control, and Computing, 1743-1750, 2011</i></p></li>

			  <li style="font-size: x-small; line-height: 130%"><p>Lucks JB, Mortimer SA, Trapnell C, Luo S, Aviran S, Schroth GP, Pachter L, Doudna JA, Arkin AP. <b> <a href="http://www.pnas.org/content/108/27/11063.long">Multiplexed RNA structure characterization with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
				<i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106501108</p><br/>
				
			  <li style="font-size: x-small; line-height: 130%"><p>Mortimer SA, Trapnell C, Aviran S, Pachter L and Lucks JB. <b> <a href="http://onlinelibrary.wiley.com/doi/10.1002/9780470559277.ch120019/abstract">SHAPE-Seq High Throughput RNA Structure Analysis</a></b> <br/>
			  <i><a href="http://onlinelibrary.wiley.com/book/10.1002/9780470559277">Current Protocols in Chemical Biology</a></i>10.1002/9780470559277.ch120019</p><br/>

			  <li style="font-size: x-small; line-height: 130%"><p>Loughrey D, Watters KE, Settle AH, and Lucks JB. <b> <a href="http://dx.doi.org/10.1093/nar/gku909">SHAPE-Seq 2.0: Systematic Optimization and Extension of High-Throughput Chemical Probing of RNA Structure with Next Generation Sequencing</a></b> <br/>
			  <i><a href="http://dx.doi.org/10.1093/nar/gku909"></a>Nucleic Acids Research</i> 10.1093/nar/gku909</p><br/>
			  </li>
			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S and Pachter L. <b> <a href="http://rnajournal.cshlp.org/content/early/2014/10/20/rna.043844.113.short ">Rational experiment design for sequencing-based RNA structure mapping</a></b> <br/>
			  <i><a href="http://rnajournal.cshlp.org/content/early/2014/10/20/rna.043844.113.short ">RNA</a> 10.1261/rna.043844.113</i></p><br/>
			  </li>
			  <li style="font-size: x-small; line-height: 130%"><p>Watters KE, Abbott TR, and Lucks JB. <b> <a href="http://dx.doi.org/10.1093/nar/gkv879">Simultaneous characterization of cellular RNA structure and function with in-cell SHAPE-Seq</a></b> <br/>
			  <i><a href="http://dx.doi.org/10.1093/nar/gkv879"></a>Nucleic Acids Research</i> 10.1093/nar/gkv879</p><br/>
			  </li>
            </ul>
        </div>
        <h2>Other Contributors</h2>
        <div class="box">
          <ul>
            <li><a href="http://rna.berkeley.edu/">Stefanie Mortimer</a></li>
            <li>Shujun Luo</li>
            <li>Gary Schroth</li>
            <li>Adam Arkin</li>
		  	  <li><a href="http://rna.berkeley.edu/">Jennifer Doudna</a></li>
          </ul>
        </div>
        <h2>Links</h2>
        <div class="box">
          <ul>
			  <li><a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a></li>
			  <li><a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a></li>
              <li><a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a></li>
              <li><a href="http://luckslab.org/">Lucks Lab</a></li>
          </ul>
        </div>
      </div>
          
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	  <table><tr><td cellpadding=7>
  	  <h1>Manual</h1><br/>
      <div id="toc">
  	    <ul>
  	    <li><a href="#prer">Prerequisites</a></li>
		<li><a href="#trim">Adapter Trimming</a></li>
  	    <li><a href="#spats">Running Spats</a></li>
		<ul>
		  <li><a href="#example">Example</a></li>
  	      <li><a href="#details">Detailed Usage</a></li>
		  <li><a href="#spats_input">Input Files</a></li>
  	  	  <li><a href="#spats_output">Output Files</a></li>
 	    </ul>
  	    </ul><br/>
  	  </div>
  	  
	  <h2 id="prer">Prerequisites</h2><br/>
	  <p>
	  Spats runs on intel-based computers running Linux or Mac OS X and 
	  that have GCC 4.0 or greater installed.  Currently spats must be built from the source code, though we recognize the need
	  for an easier installation system. Detailed instructions on building from the source are described in the <a href="installation.html">installation manual</a>. 
      </p><br/>
 	  <br/>

	  <h2 id="trim">Adapter Trimming</h2><br/>
	  <p>
		Sometimes it may be necessary to trim adapter sequences from sequencing reads prior to alignment and
        processing. For example, when SHAPE cDNAs are generated that are shorter than the sequencing read length
        used in the experiment, sequences from the required library adapters may show up in the read. Without
        trimming, these 'hybrid' adapter-cDNA sequences will not align with RNA sequences included in the spats
        rna_targets.fa index file (see below). This will result in the absence of reactivity information at the very
        3' end of molecules.
      </p>
      <br/>	
		<b>We highly recommend using the adapter trimming algorithm available via the spats v1.0.0 release.</b>
	  <p>
      </p>
 	  <br/>

  	  <h2 id="spats">Running Spats</h2><br/>
		The spats pipeline is run in two stages: adapter trimming and spats analysis. Adapter trimming is implemented in the 
		<tt>adapter_trimmer</tt> code, while spats analysis is implemented in the <tt>spats</tt> toolkit. A typical workflow is as follows:
		<p></p></br>
	   <blockquote>
		adapter_trimmer --A-b-sequence=&lt;second_adapter_sequence&gt; --A-t-sequence=&lt;first_adapter_sequence&gt; --read-len=35 R1.fastq R2.fastq RNA_targets.fa
	   </blockquote>
	   <blockquote>
		spats --num-mismatches 0 -o Output RNA_targets.fa RRRY YYYR combined_R1.fastq combined_R2.fastq
	   </blockquote>
		<p></p></br>
	    where <tt>&lt;second_adapter_sequence&gt;</tt> is the sequence of the second adapter (present at the 3' end of cDNAs), 
	    <tt>&lt;first_adapter_sequence&gt;</tt> is the 
	    sequence of the first adapter (present at the 5' end of cDNAs), <tt>R1.fastq</tt> and <tt>R2.fastq</tt> are the fastq sequencing files from the sequencing 
	    run, and <tt>RNA_targets.fa</tt> is the FASTA formatted file containing the RNA sequences under study. 
	    Spats outputs text files (in <tt>Output</tt>) containing (+) and (-) fragment counts, beta's and thetaâ€™s, for each position in each RNA in the targets file 
	    as discussed below. Note that <tt>adapter_trimmer</tt> creates the files <tt>combined_R1.fastq</tt> and <tt>combined_R2.fastq</tt> which are then fed into 
	    the spats command.
		<p>
	    </p>
	 	<br/>
		
	  	<h2 id="example">Example</h2><br/>
		Given a paired end run (2x35bp) with target sequences (in DNA alphabet) in the file 
		target_bcs.fa, "_1" reads in HCT20611_50bp_s_7_1_sequence.fq, and 
		"_2" reads in HCT20611_50bp_s_7_2_sequence.fq, you can run the mapping pipeline like this:
		<p></p><br/>
		<blockquote>
		adapter_trimmer --A-b-sequence=AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG 
		                --A-t-sequence=AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATC 
		                --read-len=35 HCT20611_50bp_s_7_1_sequence.fq HCT20611_50bp_s_7_2_sequence.fq target_bcs.fa
	    </blockquote>
	    <blockquote>
		spats --num-mismatches 0 -o HCT20611_50bp.final RRRY YYYR combined_R1.fastq combined_R2.fastq
		</blockquote>
		<p></p><br/>	
		
	  <h2 id="details">Detailed Usage</h2><br/>	

    <b><tt>adapter_trimmer</tt></b><p></p><br/>
	<blockquote>
 	   Usage: adapter_trimmer [options]* &lt;R1_seq.fastq&gt; &lt;R2_seq.fastq&gt; &lt;targets.fa&gt;
	</blockquote>  

	  <table CELLSPACING=15>
	  <tr><td VALIGN=top>
	  <b>Arguments:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  <tr><td VALIGN=top nowrap>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;R1_seq.fastq&gt;</tt>
	  </td><td VALIGN=top>
	  A fastq file containing "_1" reads from a paired-end Illumina run.
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;R2_seq.fastq&gt;</tt>
	  </td><td VALIGN=top>
	  A fastq file containing "_2" reads from a paired-end Illumina run.
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;targets.fa&gt;</tt>
	  </td><td VALIGN=top>
	  A FASTA file containing the target RNA sequences in the SHAPE experiment 
	  (excluding reactivty handles, Spats will add these automatically).
	  </td></tr>

	  <tr><td VALIGN=top>
	  <b>Options:</b>
	  </td><td VALIGN=top>
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>--trim-match &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Number of nucleotides of adapter to search for with clipper (default = 6).
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>--read-len &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Number of length of reads (ex. 35 for 2x35 bp paired end reads) (default = 35). (autodetected)
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--min-read-len &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Number of nt on the 3' end to leave after (w/ adapter) (default = 6).
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>-e, --error-rate &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Allowed error rate when clipping adapters (0.1 is 1/10 bases, etc.)
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>-q, --quality-min &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Minimum quality score to consider.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>-p, --num-threads &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	   Number of processors to use (default = 1).	
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--allow-mismatches &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Allow downstream mismatches for Spats (keep sequences that don't perfectly align).
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--set-quality &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Sets all read qualities to the given quality where ascii is (N+33).               
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--A-b-sequence &lt;string&gt;</tt>
	  </td><td VALIGN=top>
	  The sequence of A_adapter_b (which is ligated to the single-stranded cDNA products of the RT reaction). (default: 
	  AGATCGGAAGAGCACACGTCTGAACTCCAGTCAC used for Illumina TrueSeq indexed reads).	
	  </td></tr>
		
	  <tr><td VALIGN=top nowrap>
	  <tt>--A-t-sequence &lt;string&gt;</tt>
	  </td><td VALIGN=top>
	  The sequence of A_adapter_t (which contains the RT primer). (default: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCT).
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--max-handle-len &lt;N&gt;</tt>
	  </td><td VALIGN=top>
	  Number of nucleotides in the longest (+/-) handle (default = 4 for RRRY/YYYR).
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>-o/--output-dir &lt;string&gt;</tt>
	  </td><td VALIGN=top>
	  If running many simultaneously runs in the same folder, use this option to name the treated reads folder to avoid overwriting during processing.
	  </td></tr>
      
      <tr><td VALIGN=top>
	  <b>Output:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  <tr><td VALIGN=top nowrap>

	  <tr><td VALIGN=top nowrap>
	  <tt>combined_R1.fastq, combined_R2.fastq</tt>
	  </td><td VALIGN=top>
	  Fastq files containing "_1" and "_2" reads (respectively) for which all adpater sequences have been removed.
	  </td></tr>
   	  </table><br/>
	
	<b><tt>spats</tt></b><p></p><br/>
	<blockquote>
 	   Usage: spats [options]* &lt;rna_targets.fa&gt; &lt;treated_handle&gt; &lt;untreated_handle&gt;  &lt;reads1_1[,...,readsN_1]&gt; &lt;reads1_1[,...,readsN_1]&gt;
	  </blockquote>
	  <p>
	  The following is a detailed description of the options used to control
	  Spats:
	  </p><br/>

	  <table CELLSPACING=15>
	  <tr><td VALIGN=top>
	  <b>Arguments:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;rna_targets.fa&gt;</tt>
	  </td><td VALIGN=top>
	  A FASTA file containing the target RNA sequences in the SHAPE experiment 
	  (excluding reactivty handles (i.e. RRRY/YYYR), Spats will add these automatically).
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;treated_handle&gt;</tt>
	  </td><td VALIGN=top>
	  A nucleotide sequence with which all targets in the <b>treated</b> sample are tagged.  You can use IUPAC codes here, e.g. <tt>RRRY</tt>.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;untreated_handle&gt;</tt>
	  </td><td VALIGN=top>
	  A nucleotide sequence with which all targets in the <b>untreated</b> sample are tagged. You can use IUPAC codes here, e.g. <tt>YYYR</tt>.
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;reads1_1[,...,readsN_1]&gt;</tt>
	  </td><td VALIGN=top>
	  A comma separated list of FASTQ files.  These must all be "_1" reads from a paired-end Illumina run (e.g. s_1_1_sequence.txt,s_2_1_sequence.txt)
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;reads1_2[,...,readsN_2]&gt;</tt>
	  </td><td VALIGN=top>
	  A comma separated list of FASTQ files.  These must all be "_2" reads 
	  from a paired-end Illumina run (e.g. s_1_2_sequence.txt,s_2_2_sequence.txt).  
	  <b>NOTE</b>: There must be one file here for each file in the _1 list, 
	  and the files must appear in the same order.
	  </td></tr>

	  <tr><td VALIGN=top>
	  <b>Options:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  
	  <tr><td VALIGN=top nowrap>
	  <tt>-o/--output-dir &lt;output_dir_name&gt;</tt>
	  </td><td VALIGN=top>
	  The name of the directory in which you'd like Spats to write its output. Default <tt>spats_out</tt>.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--adapter-t &lt;adapter sequence&gt;</tt>
	  </td><td VALIGN=top>
	  <b>Deprecatated in v0.8.0. It is preferrable to handle adapter trimming via the <tt>adapter_trimmer</tt> script outlined above. </b> The sequence of 
	  adapter_t (which contains the RT primer).  This option is used when adapter trimming is desired. Example 
	  <tt>AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATC</tt>
	  </td></tr>
		
	  <tr><td VALIGN=top nowrap>
	  <tt>--adapter-b &lt;adapter sequence&gt;</tt>
	  </td><td VALIGN=top>
	  <b>Deprecatated in v0.8.0. It is preferrable to handle adapter trimming via the <tt>adapter_trimmer</tt> script outlined above. </b> The sequence of 
	  adapter_b (which is ligated to the single-stranded cDNA products of the RT reaction).  This option is used when adapter trimming is desired. Example    
	  <tt>AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG</tt>
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--num-mismatches &lt;0-3&gt;</tt>
	  </td><td VALIGN=top>
	  The number of mismatches you want Bowtie to allow in its alignments for each read to the target RNAs.<b>Note: this option does not work in v0.8.0 because of 
	  the algorithm implemented in <tt>adapter_trimmer</tt>.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--num-threads &lt;threads&gt;</tt>
	  </td><td VALIGN=top>
	  The number of threads you want Bowtie to use when aligning the reads to the target RNAs
	  </td></tr>
	  
	  <tr><td VALIGN=top>
	  <b>Output:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  <tr><td VALIGN=top nowrap>

	  <tr><td VALIGN=top nowrap>
	  <tt>see below</tt>
	  </td><td VALIGN=top>
	  </td></tr>
	  </table><br/>
	
	  <h2 id="spats_input">Spats Pipeline Input</h2><br/>
		<p>
		The spats pipeline (<tt>adapter_trimmer</tt> and <tt>spats</tt>) takes three files as input: the target RNA file in FASTA format, and two files containing 
		the paired-end sequencing reads from the experiment in FASTQ format. The sequecing reads for both the plus and minus channels of the
        SHAPE experiment must be contained in the files. Please refer to the references on the right hand side for
        more information on how these handles are included in library preparation and sequencing. It is critical
        that the mated reads appear in the same order in each file. The target RNA file is in FASTA format, and
        contains one entry for each RNA that is included in the experiment, e.g.
		</p>
		<br/>
		<pre>
		>target_GTAC
		GGTCGTGCCTAGCGAAGTCATAAGCTAGGGCAGTCTTTAGAGGCTGACGGCAGGAAAAAAG
		CCTACGTCTTCGGATATGGCTGAGTATCCTTGAAAGTGCCACAGTGACGAAGTCTCACTAG
		AAATGGTGAGAGTGGAACGCGGTAAACCCCTCGACCGATCCGCTTCGGCGGATCCGTACAA
		ATCGGGCTTCGGTCCGGTTC
		>target_TTAA
		GGTCGTGCCTAGCGAAGTCATAAGCTAGGGCAGTCTTTAGAGGCTGACGGCAGGAAAAAAG
		CCTACGTCTTCGGATATGGCTGAGTATCCTTGAAAGTGCCACAGTGACGAAGTCTCACTAG
		AAATGGTGAGAGTGGAACGCGGTAAACCCCTCGACCGATCCGCTTCGGCGGATCCTTAAAA
		ATCGGGCTTCGGTCCGGTTC
		...
		</pre>
		In this case, two identical RNA sequences were barcoded with two sequences (GTAC, TTAA) and run in the same
        experiment.
      <br/>
      <br/>
  <h2 id="spats_output">Spats Pipeline Output</h2><br/>
	
	The spats pipeline produces a directory (by default called <tt>spats_out</tt>) containing several output files.
	<p></p></br>
	
	<h3>De-multiplexed reads</h3>
	Spats splits the reads into the plus and minus channels according to the 
	handle sequences you've specified.  The output directory will contain FASTQ
	files named <tt>&lt;treated_handle&gt;_1.fq</tt>, <tt>&lt;treated_handle&gt;_2.fq</tt>, <tt>&lt;untreated_handle&gt;_1.fq</tt> and <tt>&lt;untreated_handle&gt;_2.fq</tt> 
	<p></p></br>
	<h3>Reactivities</h3>
		<p>
		The <tt>reactivities.out</tt> file is a tab-delimited file of values 
		providing the reactivity at each base of each target RNAs.  The example 
		below shows the first few rows of a run against wild-type RNaseP (barcoded with <tt>CTTG</tt>) and 
		tRNA (barcoded with <tt>AAAG</tt>>). Data for different targets are concatenated within the same output file.
		</p>
		<br/>
	<pre>
		sequence        rt_start  five_prime_offset  nucleotide  treated_mods  untreated_mods  beta           theta                c
		RNASeP_WT_CTTG     202            0               *          1144           4201         -              -               1.22577
		RNASeP_WT_CTTG     202            1               G          28             110          0              0               1.22577
		RNASeP_WT_CTTG     202            2               G          87             315          0.000908854    0.000883849     1.22577
		RNASeP_WT_CTTG     202            3               T          10             6            0.00553418     0.00539681      1.22577
		RNASeP_WT_CTTG     202            4               C          7              23           0.000459691    0.000446923     1.22577
		RNASeP_WT_CTTG     202            5               G          8              21           0.00146664     0.00142677      1.22577	
		...
		tRNA_AAAG          136            0               *          1267           30440        -              -               2.53892
		tRNA_AAAG          136            1               G          118            1145         0.0213467      0.0205298       2.53892
		tRNA_AAAG          136            2               G          28             123          0.00672447     0.00635241      2.53892
		tRNA_AAAG          136            3               C          17             51           0.00432867     0.0040774       2.53892
		tRNA_AAAG          136            4               C          84             319          0.0193321      0.0185458       2.53892
		tRNA_AAAG          136            5               T          95             529          0.0182941      0.0175276       2.53892
		...
	</pre>
	<table CELLSPACING=15>
	
	<tr><td VALIGN=top nowrap><tt>sequence</tt></td><td VALIGN=top>	Name of target RNA (from the input sequences FASTA file).</td></tr>
	<tr><td VALIGN=top nowrap><tt>rt_start</tt></td><td VALIGN=top>	The base index where RT priming started. For SHAPE-Seq v2.0 this is the end of the RNA. For RT 
		                                                            primers that target internal sequences this is internal to the RNA sequence.</td></tr>
	<tr><td VALIGN=top nowrap><tt>five_prime_offset</tt></td><td VALIGN=top>	Position in target RNA, 0=full length</td></tr>
	<tr><td VALIGN=top nowrap><tt>nucleotide</tt></td><td VALIGN=top>	Base at position in target, *=full length</td></tr>
	<tr><td VALIGN=top nowrap><tt>treated_mods</tt></td><td VALIGN=top>	Number of (+) fragments ending at position in target</td></tr>
	<tr><td VALIGN=top nowrap><tt>untreated_mods</tt></td><td VALIGN=top>	Number of (-) fragments ending at position in target</td></tr>
	<tr><td VALIGN=top nowrap><tt>beta</tt></td><td VALIGN=top>	Beta value at position in target</td></tr>
	<tr><td VALIGN=top nowrap><tt>theta</tt></td><td VALIGN=top>Theta value at position in target</td></tr>
	<tr><td VALIGN=top nowrap><tt>c</tt></td><td VALIGN=top>	Estimated SHAPE modifications per target</td></tr>
	</table>
	See the references, and especially <a href="http://arxiv.org/abs/1106.5061v2">Aviran, Lucks, Pachter: RNA structure characterization from chemical mapping experiments</a>, for definitions of beta, theta and c.
	</div>
  </div>
  <div id="footer">
      <table cellspacing="15" width="100%">
        <tbody>
          <tr>
            <td> This research was supported in part by
              the Miller Institute for Basic Research in Science at UC
              Berkeley, the Alfred P. Sloan Foundation, and the National Institutes of Health. </td>
            <td align="right"> Administrators: <a
                href="mailto:cole@cs.umd.edu">Cole Trapnell</a> and <a
	                  href="">Julius B. Lucks (j b lucks at cornell dot edu)</a>.
              Design by <a href="http://www.free-css-templates.com"
                title="Design by David Herreman">David Herreman</a> </td>
          </tr>
        </tbody>
      </table>
  </div>
</div>

<!-- Google analytics code -->
<!-- <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script> -->
<!-- End Google analytics code -->
</body>
</html>
