<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Spats - Read Processing and Reactivity Estimation for  SHAPE-Seq</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
	<meta name="description" content="Open source tools for transcript assembly, differential expression, and
        differential regulation for RNA-Seq">
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen">
  </head>
  <body>
    <div id="wrap">
      <div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Spats</h1>
                  </a>
                  <h2>Read Processing and Reactivity Generation for SHAPE-Seq</h2>
                </td>
                <td align="right" valign="middle"> 
                    <a href="http://bio.math.berkeley.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/UCBerkeley-seal.scaled.gif" border="0">
                    </a>&nbsp;
                    <a href="http://http://www.cheme.cornell.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/Cornell_logo.jpg" border="0">
                    </a>&nbsp;
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div id="subheader">
        <table width="100%">
          <tbody>
            <tr>
              <td> <strong>Spats</strong> processes reads and calculates SHAPE reactivities for SHAPE-Seq
              experiments on multiple RNAs. It accepts raw paired-end sequencing reads in fastq format, and a target
              sequence file containing the sequences of RNAs present in the experimental pool. Spats then performs
              read alignment to calculate distributions of read ends in the SHAPE (+) and (-) channel for each
              nucleotide in each RNA. Spats then estimates nucleotide resolution SHAPE reactivities for each RNA,
              using a model-driven maximum likelihood procedure based on a model of the reverse transcriptase
              process used in the SHAPE-Seq experiment.	
                Spats is a collaborative effort between the
				 <a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a> at UC Davis, the <a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a> at UC Berkeley, the 
                <a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a>
                 at the University of Washington, and
                <a href="http://luckslab.org/">the Lucks Lab</a>
                at Cornell University. <br>
                <br>
                Spats is provided under the OSI-approved <a
                  href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost
                  License</a>. fastx_spats_toolkit_v_0.1.0 is provided under the <a
                  href="http://en.wikipedia.org/wiki/Affero_General_Public_License">AGPLv3
                  License</a>. </td> 
              <td align="right" valign="middle"> <a
                  href="http://opensource.org"><img alt="Open Source
                    Software" src="images/osi-certified.gif" border="0"></a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div id="main">
        <div id="rightside">
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="installation.html">Installation</a></li>
              <li><a href="manual.html">Manual</a></li>
			  <!-- <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li> -->
            </ul>
          </div>
          <!-- <h2>News and updates</h2>
                    <div class="box">
                      <ul>
                        <table width="100%">
                          <tbody>
                            <tr>
                              <td>New releases and related tools will be announced
                                through the <a
                                  href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing


                                    list</b></a></td>
                            </tr>
                          </tbody>
                        </table>
                      </ul>
                    </div> -->

 		  <!-- <h2>Getting Help</h2>
 		            <div class="box">
 		              <ul>
 		                <table width="100%">
 		                  <tbody>
 		                    <tr>
 		                      <td>Questions about Cufflinks should be sent to <a
 		                          href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
 		                        Please do not email technical questions to
 		                        Cufflinks contributors directly.</td>
 		                    </tr>
 		                  </tbody>
 		                </table>
 		              </ul>
 		            </div> -->

          <a href="./downloads">
            <h2><u>Releases</u></h2>
          </a>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>spats version 0.8.0</td>
                    <td align="right">X/X/2014</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/spats-0.8.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                  <tr>
                    <td>fastx_spats_toolkit version 0.1.0</td>
                    <td align="right">X/X/2014</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/fastx_spats_toolkit-0.1.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                  <tr>
                    <td>spats version 0.7.0</td>
                    <td align="right">4/10/2012</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/spats-0.7.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>

          <h2>Related Tools</h2>
          <div class="box">
            <ul>
              <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
                Ultrafast short read alignment</li>
              <li><a href="http://hannonlab.cshl.edu/fastx_toolkit/">FastX Toolkit</a>:
                FASTQ/A short-reads pre-processing tools</li>
            </ul>
          </div>

          <h2>Publications</h2>
          <div class="box">
            <ul>

 			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Trapnell C, Lucks JB, Mortimer SA, Luo S, Schroth GP, Doudna JA, Arkin AP, Pachter L.<b> <a href="http://www.pnas.org/content/108/27/11069.long">Modeling and automation of sequencing-based characterization of RNA structure</b></a> <br/><i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106541108</p><br/></li>

		      <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Lucks JB, Pachter L.<b> <a href="http://arxiv.org/abs/1106.5061v2">RNA structure characterization from chemical mapping experiments</a></b> <br><i><a href="Aviran_Allerton_2011.pdf"></a>Proceedings of the 49th Allerton Conference on Communication, Control, and Computing, 1743-1750, 2011</i></p></li>

			  <li style="font-size: x-small; line-height: 130%"><p>Lucks JB, Mortimer SA, Trapnell C, Luo S, Aviran S, Schroth GP, Pachter L, Doudna JA, Arkin AP. <b> <a href="http://www.pnas.org/content/108/27/11063.long">Multiplexed RNA structure characterization with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
				<i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106501108</p><br/>
				
			  <li style="font-size: x-small; line-height: 130%"><p>Mortimer SA, Trapnell C, Aviran S, Pachter L and Lucks JB. <b> <a href="http://onlinelibrary.wiley.com/doi/10.1002/9780470559277.ch120019/abstract">SHAPE-Seq High Throughput RNA Structure Analysis</a></b> <br/>
			  <i><a href="http://onlinelibrary.wiley.com/book/10.1002/9780470559277">Current Protocols in Chemical Biology</a></i>10.1002/9780470559277.ch120019</p><br/>

			  <li style="font-size: x-small; line-height: 130%"><p>Loughrey D, Watters KE, Settle AH, and Lucks JB. <b> <a href="XXX">SHAPE-Seq 2.0: Systematic Optimization and Extension of High-Throughput Chemical Probing of RNA Structure with Next Generation Sequencing</a></b> <br/>
			  <i><a href=""></a></i></p><br/>
			  </li>
            </ul>
          </div>
          <h2>Other Contributors</h2>
          <div class="box">
            <ul>
              <li><a href="http://rna.berkeley.edu/">Stefanie Mortimer</a></li>
              <li>Shujun Luo</li>
              <li>Gary Schroth</li>
              <li>Adam Arkin</li>
		  	  <li><a href="http://rna.berkeley.edu/">Jennifer Doudna</a></li>
            </ul>
          </div>
          <h2>Links</h2>
          <div class="box">
            <ul>
			  <li><a href="http://bme.ucdavis.edu/aviranlab/">Aviran Lab</a></li>
			  <li><a href="http://cole-trapnell-lab.github.io">Trapnell Lab</a></li>
              <li><a href="http://math.berkeley.edu/~lpachter/">Pachter Lab</a></li>
              <li><a href="http://luckslab.org/">Lucks Lab</a></li>
            </ul>
          </div>
        </div>

        <!-- End of "rightside" -->
        <div id="leftside">
          <h2>0.8.0 release - X/X/2014</h2>
          <p> Spats v 0.8.0 is now available for download. This release was made to support SHAPE-Seq v2.0, an experimental
	          upgrade to the SHAPE-Seq experimental workflow that removes the need to use internal RT primer sites or additional
	          flanking RNA sequences to act as RT priming sites. SHAPE-Seq 2.0 ligates a linker after RNA modification, which
	          is then used to prime the RT reaction. Details about SHAPE-Seq 2.0 can be found in <a href="">Loughrey, Watters et. al. (2014)</a>.
	
	      <p> The major difference between v0.7.0 and v0.8.0 is an updated
	          adapter trimming algorithm. This now requires installation of a modified fastx toolkit (fastx_spats_toolkit v0.1.0)
	          which can be found here <a href="./downloads/fastx_spats_toolkit-0.1.0.tar.gz"
                  onclick="javascript:
                  pageTracker._trackPageview('/downloads/cufflinks_source');
                  ">here</a>. A new python script apdater_trimmer.py was made to implement the
	          trimming algorithm, which should be used as a pre-process step before calling spats. Details of how to use the
	          new implementation can be found in the <a href="manual.html">manual</a>.
	      
	      <p> We have also greatly expanded the <a href="installation.html">installation</a> instructions.
	
	      <p> A special thanks of Gordon Assaf for assisting with the coding and release of fastx_spats_toolkit_v0.1.0 .
	
		  <p>This software is a work in progress
	          - it is a beta release, and new features will continue to be added over time. To suggest a feature,
	          report a bug, or report a website error please <a href="mailto:spats.shape@gmail.com">email us</a>
	      <p></p></br>
	    </div>
        <div id="leftside">
          <h2>0.7.0 release - 4/10/2012</h2>
          <p> The first public release of Spats is now available for download. Spats is a program for the analysis
          of sequencing information to determine SHAPE reactivities in a SHAPE-Seq experiment. It runs on Linux and
          OS X. 
          <p>Spats takes as input files containing paired-end reads in <b>fastq</b> format, a file containing
          the sequences of RNA molecules contained in the experimental pool in FASTA format, and reports
          reactivities for each nucleotide of each RNA in a flat file format. 
          <p>This software is a work in progress
          - it is a beta release, and new features will continue to be added over time. To suggest a feature or
          report a bug, please <a href="mailto:spats.shape@gmail.com">email us</a>
          </p>
        </div>
      </div>
      <div id="footer">
        <table cellspacing="15" width="100%">
          <tbody>
            <tr>
              <td> This research was supported in part by
                the Miller Institute for Basic Research in Science at UC
                Berkeley, the Alfred P. Sloan Foundation, and the National Institutes of Health. </td>
              <td align="right"> Administrators: <a
                  href="mailto:cole@cs.umd.edu">Cole Trapnell</a> and <a
	                  href="">Julius B. Lucks (j b lucks at cornell dot edu)</a>.
                Design by <a href="http://www.free-css-templates.com"
                  title="Design by David Herreman">David Herreman</a> </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <!-- Google analytics code -->
    <!-- <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
    <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script> -->
    <!-- End google analytics code -->
  </body>
</html>
