<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Spats - Getting Started</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"></meta>
	<meta name="description" content="A brief guide to using Spats for RNA structure determination with SHAPE-Seq"></meta>
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen"></link>
  </head>
  <body>
    <div id="wrap">
     	<div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Spats</h1>
                  </a>
                  <h2>Read Processing and Reactivity Generation for SHAPE-Seq</h2>
                </td>
                <td align="right" valign="middle"> 
                    <a href="http://bio.math.berkeley.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/UCBerkeley-seal.scaled.gif" border="0">
                    </a>&nbsp;
                    <a href="http://http://www.cheme.cornell.edu/">
                    <img style="vertical-align:middle;padding-top:4px"
                      src="images/Cornell_logo.jpg" border="0">
                    </a>&nbsp;
                </td>
              </tr>
            </tbody>
          </table>
        </div>

      <div id="main">
        <div id="subheader">
	        <table width="100%">
	          <tbody>
	            <tr>
	              <td> <strong>Spats</strong> processes reads and calculates SHAPE reactivities for SHAPE-Seq
	              experiments on multiple RNAs. It accepts raw paired-end sequencing reads in fastq format, and a target
	              sequence file containing the sequences of RNAs present in the experimental pool. Spats then performs
	              read alignment to calculate distributions of read ends in the SHAPE (+) and (-) channel for each
	              nucleotide in each RNA. Spats then estimates nucleotide resolution SHAPE reactivities for each RNA,
	              using a model-driven maximum likelihood procedure based on a model of the reverse transcriptase
	              process used in the SHAPE-Seq experiment.	
	                <br>
	                Spats is a collaborative effort between the <a
	                  href="http://bio.math.berkeley.edu/">Laboratory for
	                  Mathematical and Computational Biology</a>, led by
	                Lior Pachter at UC Berkeley, 
	                <a href="http://www.broadinstitute.org/~cole/">Cole Trapnell</a>
	                 at the Department of Stem Cell and Regenerative Biology at Harvard University, and
	                <a href="http://luckslab.org/">the Lucks Lab</a>
	                at Cornell University. <br>
	                <br>
	                Spats is provided under the OSI-approved <a
	                  href="http://en.wikipedia.org/wiki/Boost_Software_License">Boost
	                  License</a> </td>
	              <td align="right" valign="middle"> <a
	                  href="http://opensource.org"><img alt="Open Source
	                    Software" src="images/osi-certified.gif" border="0"></a>
	              </td>
	            </tr>
	          </tbody>
	        </table>
	      </div>
	
		<div id="rightside">
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="installation.html">Installation</a></li>
              <li><a href="manual.html">Manual</a></li>
			  <!-- <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li> -->
            </ul>
          </div>
          <!-- <h2>News and updates</h2>
                    <div class="box">
                      <ul>
                        <table width="100%">
                          <tbody>
                            <tr>
                              <td>New releases and related tools will be announced
                                through the <a
                                  href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing


                                    list</b></a></td>
                            </tr>
                          </tbody>
                        </table>
                      </ul>
                    </div> -->

 		  <!-- <h2>Getting Help</h2>
 		            <div class="box">
 		              <ul>
 		                <table width="100%">
 		                  <tbody>
 		                    <tr>
 		                      <td>Questions about Spats should be sent to <a
 		                          href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
 		                        Please do not email technical questions to
 		                        Spats contributors directly.</td>
 		                    </tr>
 		                  </tbody>
 		                </table>
 		              </ul>
 		            </div> -->

          <a href="./downloads">
            <h2><u>Releases</u></h2>
          </a>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>version 0.7.0</td>
                    <td align="right">X/X/2012</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/spats-0.7.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>

          <h2>Related Tools</h2>
          <div class="box">
            <ul>
              <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
                Ultrafast short read alignment</li>
              <li><a href="http://hannonlab.cshl.edu/fastx_toolkit/">FastX Toolkit</a>:
                FASTQ/A short-reads pre-processing tools</li>
            </ul>
          </div>

          <h2>Publications</h2>
          <div class="box">
            <ul>

 			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Trapnell C, Lucks JB, Mortimer SA, Luo S, Schroth GP, Doudna JA, Arkin AP, Pachter L.<b> <a href="http://www.pnas.org/content/108/27/11069.long">Modeling and automation of sequencing-based characterization of RNA structure</b></a> <br/><i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106541108</p><br/></li>

		      <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Lucks JB, Pachter L.<b> <a href="http://arxiv.org/abs/1106.5061v2">RNA structure characterization from chemical mapping experiments</a></b> <br><i><a href="http://arxiv.org/">arXiv</a></i> arXiv:1106.5061v2</p></li>

			  <li style="font-size: x-small; line-height: 130%"><p>Lucks JB, Mortimer SA, Trapnell C, Luo S, Aviran S, Schroth GP, Pachter L, Doudna JA, Arkin AP. <b> <a href="http://www.pnas.org/content/108/27/11063.long">Multiplexed RNA structure characterization with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
				<i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106501108</p><br/>
			  </li>

            </ul>
          </div>
          <h2>Contributors</h2>
          <div class="box">
            <ul>
		  	  <li>Cole Trapnell</li>
              <li>Sharon Aviran</li>
              <li><a href="http://luckslab.org/">Julius Lucks</a></li>
              <li>Stefanie Mortimer</li>
              <li>Shujun Luo</li>
              <li>Gary Schroth</li>
              <li>Adam Arkin</li>
		  	  <li>Jennifer Doudna</li>
              <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior Pachter</a></li>
            </ul>
          </div>
          <h2>Links</h2>
          <div class="box">
            <ul>
              <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
              <li><a href="http://luckslab.org/">Lucks Lab</a></li>
            </ul>
          </div>
        </div>

        <!-- End of "rightside" -->
        <div id="leftside">
          <table>
            <tbody>
              <tr>
                <td>
					<h3 id="bowtie">Installing Bowtie (Required)</h3> 
					Please install <a href="http://bowtie-bio.sourceforge.net">Bowtie</a> version 0.12.7.  Spats is not currently compatible with Bowtie 2. You will need the following Bowtie programs in your <tt>PATH</tt>:
		            <ul>
		        	      <li><tt>bowtie</tt></li>
		        	      <li><tt>bowtie-build</tt></li>
						  <li><tt>bowtie-inspect</tt></li>
		        	</ul>
					<br/>
				  <h3 id="fastx">Installing FASTX and LibGTextUtils (Optional)</h3>  
				     If you would like to include adapter trimming from reads, you will need to install FASTX and
                     LibGTextUtils, developed by members of Greg Hannon's lab, as described <a
                     href="http://hannonlab.cshl.edu/fastx_toolkit/">here</a>. The following tools must be in your
                     PATH:
					<ul>
		        	      <li><tt>fastx-clipper</tt></li>
		        	</ul> 
					<br/>
                  <h3 id="boost">Installing Boost (Required)</h3>
                  <br/>
                  <ol>
                    <li><a href="http://www.boost.org/users/download/">Download</a>
                      Boost and the <tt>bjam</tt> build engine.</li>
                    <li>Unpack bjam with the command
	                    <blockquote>
		                 ./bootstrap.sh
		                </blockquote>
	                </li>
                    <li>Add bjam to your PATH.</li>
                    <li>Unpack the Boost tarball and <tt>cd</tt> to the
                      Boost source directory. This directory is called
                      the BOOST_ROOT in some Boost installation
                      instructions.</li>
                    <li>Build Boost. Note that you can specify where to
                      put Boost with the --prefix option. The default
                      Boost installation directory is <tt>/usr/local</tt>.
                      Take note of the boost installation directory,
                      because you will need to tell the Spats
                      installer where to find Boost later on.
                      <p>If you are on Mac OS X, type (all on one line):
                        <br/>
                      </p>
                      <blockquote>bjam
                        --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                        --toolset=darwin --architecture=x86
                        --address_model=32_64 --link=static
                        --runtime-link=static --layout=versioned --stage=install</blockquote>
                    </li>
                    <p>If you are on a 32-bit Linux system, type (all on
                      one line): <br/>
                    </p>
                    <blockquote>bjam
                      --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                      --toolset=gcc --architecture=x86 --address_model=32
                      --link=static --runtime-link=static --stage=install</blockquote>
                    <p>If you are on a 64-bit Linux system, type (all on
                      one line): <br/>
                    </p>
                    <blockquote>bjam
                      --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                      --toolset=gcc --architecture=x86 --address_model=64
                      --link=static --runtime-link=static --stage=install</blockquote>
                  </ol>
                  <br/>
                 
                  <h3 id="boost">Building Spats (Required)</h3>
                  <br/>
                  <ol>
                    <li>Unpack the Spats source tarball:
                      <blockquote>tar zxvf spats-0.7.0.tar.gz</blockquote>
                    </li>
                    <li>Change to the Spats directory:
                      <blockquote>cd spats-0.7.0</blockquote>
                    </li>
                    <li>Configure Spats. If Boost is installed
                      somewhere other than <tt>/usr/local</tt>, you
                      will need to tell the installer where to find it
                      using the <tt>--with-boost</tt> option. Specify
                      where to install Spats using the <tt>--prefix</tt>
                      option.
                      <blockquote>./configure
                        --prefix=/path/to/spats/install
                        --with-boost=/path/to/boost</blockquote>
                      If you see any errors during configuration, verify
                      that you are using Boost version 1.38 or higher,
                      and that the directory you specified via <tt>--with-boost</tt>
                      contains the boost header files and libraries. See
                      the Boost <a href="">Getting started</a> page for
                      more details.  </li>
                    <li>Finally, make and install Spats.
                      <blockquote>make</blockquote>
                      <blockquote>make install</blockquote>
                    </li>
                  </ol>
                  <!-- <h2 id="test">Testing the installation</h2>
                  <br/>
                  <ol>
                    <li>Download the <a href="downloads/test_data.sam">test
                        data</a> </li>
                    <li>In the directory where you placed the test file,
                      type:<br/>
                      <blockquote>cufflinks ./test_data.sam</blockquote>
                      <br/>
                      You should see the following output:
                      <pre>[bam_header_read] EOF marker is absent. The input is probably truncated.
[bam_header_read] invalid BAM binary header (this is not a BAM file).
File ./test_data.sam doesn't appear to be a valid BAM file, trying SAM...
[13:23:15] Inspecting reads and determining fragment length distribution.
&gt; Processed 1 loci.                            [*************************] 100%
Warning: Using default Gaussian distribution due to insufficient paired-end reads in open ranges.  
It is recommended that correct paramaters (--frag-len-mean and --frag-len-std-dev) be provided.
&gt; Map Properties:
&gt;       Total Map Mass: 102.50
&gt;       Read Type: 75bp x 75bp
&gt;       Fragment Length Distribution: Truncated Gaussian (default)
&gt;                     Estimated Mean: 200
&gt;                  Estimated Std Dev: 80
[13:23:15] Assembling transcripts and estimating abundances.
&gt; Processed 1 loci.                            [*************************] 100%
</pre>
                    </li>
                    <li>Verify that the file <tt>transcripts.gtf</tt>
                      is in the current directory and looks like this
                      (your file will have GTF attributes, omitted here
                      for clarity)
                      <pre>test_chromosome Spats       exon    53      250     1000    +       . 
test_chromosome Spats       exon    351     400     1000    +       . 
test_chromosome Spats       exon    501     550     1000    +       .
					
				</pre>
                    </li>
                  </ol>
                  <br/> -->
                 
                  
                  <div id="footer">
                    <table cellspacing="15" width="100%">
                      <tbody>
                        <tr>
                          	<td> This research was supported in part by
				                the Miller Institute for Basic Research in Science at UC
				                Berkeley. </td>
                          <td align="right"> Administrator: <a
                              href="mailto:cole@cs.umd.edu">Cole
                              Trapnell</a>. Design by <a
                              href="http://www.free-css-templates.com"
                              title="Design by David Herreman">David
                              Herreman</a> </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>

                  <!-- Google analytics code -->
                  <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
                  <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
                  <!-- End google analytics code --> </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>
</html>
