<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <title>Spats - Getting Started</title>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8"></meta>
	<meta name="description" content="A brief guide to using Spats for RNA structure determination with SHAPE-Seq"></meta>
    <link rel="stylesheet" type="text/css" href="css/style.css"
      media="screen"></link>
  </head>
  <body>
    <div id="wrap">
      <div id="top">
        <div class="lefts">
          <table cellpadding="2" width="100%">
            <tbody>
              <tr>
                <td> <a href="./index.html">
                    <h1>Spats</h1>
                  </a>
                  <h2>RNA structure determination with SHAPE-Seq</h2>
                </td>
                <!-- <td align="right" valign="middle"> 
                                    <a href="http://www.mcb.berkeley.edu/">
                                      <img style="vertical-align:middle;padding-top:4px" 
                                        border=0 src="images/UCBerkeley-seal.scaled.gif">
                                    </a>&nbsp;
                                    <a href="http://genomics.jhu.edu/">
                                    <img style="vertical-align:middle;padding-top:4px"
                                      src="images/JHU-seal.gif" border="0">
                                    </a>&nbsp;
                                  <a href="http://www.cbcb.umd.edu/"><img
                                      style="vertical-align:middle;padding-top:4px"
                                      src="images/cbcb_logo.gif" alt=""></a>&nbsp;&nbsp;

                                </td> -->
              </tr>
            </tbody>
          </table>
        </div>
      </div>

      <div id="main">
        <div id="rightside">
          <h2>Site Map</h2>
          <div class="box">
            <ul>
              <li><a href="index.html">Home</a></li>
              <li><a href="tutorial.html">Getting started</a></li>
              <li><a href="manual.html">Manual</a></li>
			  <!-- <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li> -->
            </ul>
          </div>
          <!-- <h2>News and updates</h2>
                    <div class="box">
                      <ul>
                        <table width="100%">
                          <tbody>
                            <tr>
                              <td>New releases and related tools will be announced
                                through the <a
                                  href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing


                                    list</b></a></td>
                            </tr>
                          </tbody>
                        </table>
                      </ul>
                    </div> -->
         
 		  <!-- <h2>Getting Help</h2>
 		            <div class="box">
 		              <ul>
 		                <table width="100%">
 		                  <tbody>
 		                    <tr>
 		                      <td>Questions about Cufflinks should be sent to <a
 		                          href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
 		                        Please do not email technical questions to
 		                        Cufflinks contributors directly.</td>
 		                    </tr>
 		                  </tbody>
 		                </table>
 		              </ul>
 		            </div> -->
 		  
          <a href="./downloads">
            <h2><u>Releases</u></h2>
          </a>
          <div class="box">
            <ul>
              <table width="100%">
                <tbody>
                  <tr>
                    <td>version 0.7.0</td>
                    <td align="right">X/X/2012</td>
                  </tr>
                  <tr>
                    <td><a href="./downloads/spats-0.7.0.tar.gz"
                        onclick="javascript:
                        pageTracker._trackPageview('/downloads/cufflinks_source');
                        ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                  </tr>
                </tbody>
              </table>
            </ul>
          </div>

          <h2>Related Tools</h2>
          <div class="box">
            <ul>
              <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
                Ultrafast short read alignment</li>
            </ul>
          </div>

          <!-- <h2>Publications</h2>
          <div class="box">
            <ul>
			  <li style="font-size: x-small; line-height: 130%"><p>Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L.<b> <a href="http://dx.doi.org/10.1038/nbt.1621">Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</a></b> <br/>
				<i><a href="http://www.nature.com/nbt">Nature Biotechnology</a></i> doi:10.1038/nbt.1621</p><br/>
			  </li>
			
 			  <li style="font-size: x-small; line-height: 130%"><p>Roberts A, Trapnell C, Donaghey J, Rinn JL, Pachter L.<b> <a href="http://genomebiology.com/2011/12/3/R22/abstract">Improving RNA-Seq expression estimates by correcting for fragment bias</b></a> <br/><i><a href="http://www.genomebiology.com">Genome Biology</a></i> doi:10.1186/gb-2011-12-3-r22</p><br/></li>


		      <li style="font-size: x-small; line-height: 130%"><p>Roberts A, Pimentel H, Trapnell C, Pachter L.<b> <a href="http://bioinformatics.oxfordjournals.org/content/early/2011/06/21/bioinformatics.btr355.abstract">Identification of novel transcripts in annotated genomes using RNA-Seq</a></b> <br><i><a href="http://bioinformatics.oxfordjournals.org/">Bioinformatics</a></i> doi:10.1093/bioinformatics/btr355</p></li>
		
            </ul>
          </div> -->
          <h2>Contributors</h2>
          <div class="box">
            <ul>
              <li>Julius Lucks</li>
              <li>Stefanie Mortimer</li>
		  	  <li>Cole Trapnell</li>
              <li>Sharon Aviran</li>
              <li>Shujun Luo</li>
              <li>Gary Schroth</li>
              <li>Adam Arkin</li>
		  	<li>Jennifer Doudna</li>
              <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior Pachter</a></li>
            </ul>
          </div>
          <h2>Links</h2>
          <div class="box">
            <ul>
              <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
              <li><a href="http://www.cbcb.umd.edu/">UMD CBCB</a></li>
              <li><a href="http://woldlab.caltech.edu/">Wold Lab</a></li>
            </ul>
          </div>
        </div>
        <!-- End of "rightside" -->
        <div id="leftside">
          <table>
            <tbody>
              <tr>
                <td>
                  <h1> Getting started</h1>
                  <br/>
                  <div id="toc">
                    <ul>
                      <ul>
                        <li><a href="#inst">Install quick-start</a></li>
                      </ul>
                   
                    <br/>
                  </div>
                  <br/>
                  <h2 id="inst">Install quick-start</h2>
					You will need the following Bowtie programs in your <tt>PATH</tt>:
		            <ul>
		        	      <li><tt>bowtie</tt></li>
		        	      <li><tt>bowtie-build</tt></li>
						  <li><tt>bowtie-inspect</tt></li>
		        	</ul>
				  <h3 id="boost">Installing FASTX</h3>                  

                  <h3 id="boost">Installing Boost</h3>
                  <br/>
                  <ol>
                    <li><a href="http://www.boost.org/users/download/">Download</a>
                      Boost and the <tt>bjam</tt> build engine.</li>
                    <li>Unpack bjam and add it to your PATH.</li>
                    <li>Unpack the Boost tarball and <tt>cd</tt> to the
                      Boost source directory. This directory is called
                      the BOOST_ROOT in some Boost installation
                      instructions.</li>
                    <li>Build Boost. Note that you can specify where to
                      put Boost with the --prefix option. The default
                      Boost installation directory is <tt>/usr/local</tt>.
                      Take note of the boost installation directory,
                      because you will need to tell the Cufflinks
                      installer where to find Boost later on.
                      <p>If you are on Mac OS X, type (all on one line):
                        <br/>
                      </p>
                      <blockquote>bjam
                        --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                        --toolset=darwin architecture=x86
                        address_model=32_64 link=static
                        runtime-link=static --layout=versioned stage
                        install</blockquote>
                    </li>
                    <p>If you are on a 32-bit Linux system, type (all on
                      one line): <br/>
                    </p>
                    <blockquote>bjam
                      --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                      --toolset=gcc architecture=x86 address_model=32
                      link=static runtime-link=static stage install</blockquote>
                    <p>If you are on a 64-bit Linux system, type (all on
                      one line): <br/>
                    </p>
                    <blockquote>bjam
                      --prefix=&lt;YOUR_BOOST_INSTALL_DIRECTORY&gt;
                      --toolset=gcc architecture=x86 address_model=64
                      link=static runtime-link=static stage install</blockquote>
                  </ol>
                  <br/>
                 
                  <h3 id="boost">Building Spats</h3>
                  <br/>
                  <ol>
                    <li>Unpack the Spate source tarball:
                      <blockquote>tar zxvf spats-0.7.0.tar.gz</blockquote>
                    </li>
                    <li>Change to the Spats directory:
                      <blockquote>cd cufflinks-0.7.0</blockquote>
                    </li>
                    <li>Configure Spats. If Boost is installed
                      somewhere other than <tt>/usr/local</tt>, you
                      will need to tell the installer where to find it
                      using the <tt>--with-boost</tt> option. Specify
                      where to install Spats using the <tt>--prefix</tt>
                      option.
                      <blockquote>./configure
                        --prefix=/path/to/cufflinks/install
                        --with-boost=/path/to/boost</blockquote>
                      If you see any errors during configuration, verify
                      that you are using Boost version 1.38 or higher,
                      and that the directory you specified via <tt>--with-boost</tt>
                      contains the boost header files and libraries. See
                      the Boost <a href="">Getting started</a> page for
                      more details. If you copied the SAM tools binaries
                      to someplace other than <tt>/usr/local/</tt>, you
                      may need to supply the <tt>--with-bam</tt>
                      configuration option. </li>
                    <li>Finally, make and install Spats.
                      <blockquote>make</blockquote>
                      <blockquote>make install</blockquote>
                    </li>
                  </ol>
                  <!-- <h2 id="test">Testing the installation</h2>
                  <br/>
                  <ol>
                    <li>Download the <a href="downloads/test_data.sam">test
                        data</a> </li>
                    <li>In the directory where you placed the test file,
                      type:<br/>
                      <blockquote>cufflinks ./test_data.sam</blockquote>
                      <br/>
                      You should see the following output:
                      <pre>[bam_header_read] EOF marker is absent. The input is probably truncated.
[bam_header_read] invalid BAM binary header (this is not a BAM file).
File ./test_data.sam doesn't appear to be a valid BAM file, trying SAM...
[13:23:15] Inspecting reads and determining fragment length distribution.
&gt; Processed 1 loci.                            [*************************] 100%
Warning: Using default Gaussian distribution due to insufficient paired-end reads in open ranges.  
It is recommended that correct paramaters (--frag-len-mean and --frag-len-std-dev) be provided.
&gt; Map Properties:
&gt;       Total Map Mass: 102.50
&gt;       Read Type: 75bp x 75bp
&gt;       Fragment Length Distribution: Truncated Gaussian (default)
&gt;                     Estimated Mean: 200
&gt;                  Estimated Std Dev: 80
[13:23:15] Assembling transcripts and estimating abundances.
&gt; Processed 1 loci.                            [*************************] 100%
</pre>
                    </li>
                    <li>Verify that the file <tt>transcripts.gtf</tt>
                      is in the current directory and looks like this
                      (your file will have GTF attributes, omitted here
                      for clarity)
                      <pre>test_chromosome Cufflinks       exon    53      250     1000    +       . 
test_chromosome Cufflinks       exon    351     400     1000    +       . 
test_chromosome Cufflinks       exon    501     550     1000    +       .
					
				</pre>
                    </li>
                  </ol>
                  <br/> -->
                 
                  
                  <div id="footer">
                    <table cellspacing="15" width="100%">
                      <tbody>
                        <tr>
                          <td> This research was supported in part by
                            NIH grants R01-LM06845 and R01-GM083873, NSF
                            grant CCF-0347992 and the Miller Institute
                            for Basic Research in Science at UC
                            Berkeley. </td>
                          <td align="right"> Administrator: <a
                              href="mailto:cole@cs.umd.edu">Cole
                              Trapnell</a>. Design by <a
                              href="http://www.free-css-templates.com"
                              title="Design by David Herreman">David
                              Herreman</a> </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>

                  <!-- Google analytics code -->
                  <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
                  <script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script>
                  <!-- End google analytics code --> </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </body>
</html>
