<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>Spats</title>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<link rel="stylesheet" type="text/css" href="css/style.css" media="screen" />
</head>
<body>
<div id="wrap">
  <div id="top">
    <div class="lefts">
    <table width="100%" cellpadding="2">
      <tr><td>
        <a href="./index.html"><h1>Spats</h1></a>
        <h2>RNA structure determination with SHAPE-Seq</h2>
      </td><td align="right" valign="middle">
        <h1><a href="http://www.cbcb.umd.edu/"><img border=0 src="images/cbcblogo.gif"></img></a>&nbsp;&nbsp;</h1>
      </td></tr>
    </table>
    </div>
  </div>
  <div id="subheader">
  <table width="100%"><tr>
  <td>
  		<strong>Please Note</strong> If you have questions 
		  about how to use Spats or would like more information about 
		  the software, please email <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>.
	</td><td align=right valign=middle>
  </td></tr>
  </table>
  </div>
  <div id="main">
	
		<div id="rightside">
        <h2>Site Map</h2>
        <div class="box">
          <ul>
            <li><a href="index.html">Home</a></li>
            <li><a href="tutorial.html">Installation</a></li>
            <li><a href="manual.html">Manual</a></li>
			  <!-- <li><a href="http://www.nature.com/nprot/journal/v7/n3/full/nprot.2012.016.html">Protocol</a></li> -->
          </ul>
        </div>
        <!-- <h2>News and updates</h2>
                  <div class="box">
                    <ul>
                      <table width="100%">
                        <tbody>
                          <tr>
                            <td>New releases and related tools will be announced
                              through the <a
                                href="https://lists.sourceforge.net/lists/listinfo/bowtie-bio-announce"><b>mailing


                                  list</b></a></td>
                          </tr>
                        </tbody>
                      </table>
                    </ul>
                  </div> -->

		  <!-- <h2>Getting Help</h2>
		            <div class="box">
		              <ul>
		                <table width="100%">
		                  <tbody>
		                    <tr>
		                      <td>Questions about Cufflinks should be sent to <a
		                          href="mailto:tophat.cufflinks@gmail.com"><b>tophat.cufflinks@gmail.com</b></a>.
		                        Please do not email technical questions to
		                        Cufflinks contributors directly.</td>
		                    </tr>
		                  </tbody>
		                </table>
		              </ul>
		            </div> -->

        <a href="./downloads">
          <h2><u>Releases</u></h2>
        </a>
        <div class="box">
          <ul>
            <table width="100%">
              <tbody>
                <tr>
                  <td>version 0.7.0</td>
                  <td align="right">X/X/2012</td>
                </tr>
                <tr>
                  <td><a href="./downloads/spats-0.7.0.tar.gz"
                      onclick="javascript:
                      pageTracker._trackPageview('/downloads/cufflinks_source');
                      ">&nbsp;&nbsp;&nbsp;Source code</a></td>
                </tr>
              </tbody>
            </table>
          </ul>
        </div>

        <h2>Related Tools</h2>
        <div class="box">
          <ul>
            <li><a href="http://bowtie.cbcb.umd.edu">Bowtie</a>:
              Ultrafast short read alignment</li>
            <li><a href="http://hannonlab.cshl.edu/fastx_toolkit/">FastX Toolkit</a>:
              FASTQ/A short-reads pre-processing tools</li>
          </ul>
        </div>

        <h2>Publications</h2>
        <div class="box">
          <ul>

			  <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Trapnell C, Lucks JB, Mortimer SA, Luo S, Schroth GP, Doudna JA, Arkin AP, Pachter L.<b> <a href="http://www.pnas.org/content/108/27/11069.long">Modeling and automation of sequencing-based characterization of RNA structure</b></a> <br/><i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106541108</p><br/></li>

		      <li style="font-size: x-small; line-height: 130%"><p>Aviran S, Lucks JB, Pachter L.<b> <a href="http://arxiv.org/abs/1106.5061v2">RNA structure characterization from chemical mapping experiments</a></b> <br><i><a href="http://arxiv.org/">arXiv</a></i> arXiv:1106.5061v2</p></li>

			  <li style="font-size: x-small; line-height: 130%"><p>Lucks JB, Mortimer SA, Trapnell C, Luo S, Aviran S, Schroth GP, Pachter L, Doudna JA, Arkin AP. <b> <a href="http://www.pnas.org/content/108/27/11063.long">Multiplexed RNA structure characterization with selective 2'-hydroxyl acylation analyzed by primer extension sequencing (SHAPE-Seq)</a></b> <br/>
				<i><a href="http://www.pnas.org">PNAS</a></i> 10.1073/pnas.1106501108</p><br/>
			  </li>

          </ul>
        </div>
        <h2>Contributors</h2>
        <div class="box">
          <ul>
		  	  <li>Cole Trapnell</li>
            <li>Sharon Aviran</li>
            <li><a href="http://luckslab.org/">Julius Lucks</a></li>
            <li>Stefanie Mortimer</li>
            <li>Shujun Luo</li>
            <li>Gary Schroth</li>
            <li>Adam Arkin</li>
		  	  <li>Jennifer Doudna</li>
            <li><a href="http://www.math.berkeley.edu/%7Elpachter/">Lior Pachter</a></li>
          </ul>
        </div>
        <h2>Links</h2>
        <div class="box">
          <ul>
            <li><a href="http://bio.math.berkeley.edu/">Berkeley LMCB</a></li>
            <li><a href="http://luckslab.org/">Lucks Lab</a></li>
          </ul>
        </div>
      </div>      
          
    </div> <!-- End of "rightside" -->
    <div id="leftside">
  	  <table><tr><td cellpadding=7>
  	  <h1>Manual</h1><br/>
      <div id="toc">
  	    <ul>
  	    <li><a href="#prer">Prerequisites</a></li>
  	    <li><a href="#cufflinks">Running Spats</a></li>
		<ul>
  	      <li><a href="#cufflinks_input">Input Files</a></li>
  	  	  <li><a href="#cufflinks_output">Output Files</a></li>
 	    </ul>
  	    </ul><br/>
  	  </div>
  	  
	  <h2 id="prer">Prerequisites</h2><br/>
	  <p>
	  Spats runs on intel-based computers running Linux or Mac OS X and 
	  that have GCC 4.0 or greater installed.  You can install pre-compiled 
	  binaries or build Spats from the source code.
	  If you wish to build Spats yourself, you will need to install several required pieces of software as described <a href="installation.html">here</a>. 
      </p><br/>
 	  <br/>
  	  <h2 id="spats">Running Spats</h2><br/>
		Run <tt>spats</tt> from the command line as follows:
	  <blockquote>
 	   Usage: spats [options]* &lt;rna_targets.fa&gt; &lt;treated_handle&gt; &lt;untreated_handle&gt;  &lt;reads1_1[,...,readsN_1]&gt; &lt;reads1_1[,...,readsN_1]&gt;
	  </blockquote>
	  <p>
	  The following is a detailed description of the options used to control
	  Spats:
	  </p><br/>

	  <table CELLSPACING=15>
	  <tr><td VALIGN=top>
	  <b>Arguments:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;rna_targets.fa&gt;</tt>
	  </td><td VALIGN=top>
	  A FASTA file containing the target RNA sequences in the SHAPE experiment 
	  (excluding reactivty handles, Spats will add these automatically)
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;treated_handle&gt;</tt>
	  </td><td VALIGN=top>
	  A nucleotide sequence with which all targets in the <b>treated</b> sample are tagged.  You can use IUPAC codes here, e.g. <tt>RRRY</tt>
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;untreated_handle&gt;</tt>
	  </td><td VALIGN=top>
	  A nucleotide sequence with which all targets in the <b>untreated</b> sample are tagged. You can use IUPAC codes here, e.g. <tt>YYYR</tt>
	  </td></tr>

	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;reads1_1[,...,readsN_1]&gt;</tt>
	  </td><td VALIGN=top>
	  A comma separated list of FASTQ files.  These must all be "_1" reads from a paired-end Illumina run (e.g. s_1_1_sequence.txt,s_2_1_sequence.txt)
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>&lt;reads1_2[,...,readsN_2]&gt;</tt>
	  </td><td VALIGN=top>
	  A comma separated list of FASTQ files.  These must all be "_2" reads 
	  from a paired-end Illumina run (e.g. s_1_2_sequence.txt,s_2_2_sequence.txt).  
	  <b>NOTE</b>: There must be one file here for each file in the _1 list, 
	  and the files must appear in the same order.
	  </td></tr>

	  <tr><td VALIGN=top>
	  <b>Options:</b>
	  </td><td VALIGN=top>
	  </td></tr>
	  
	
	  <tr><td VALIGN=top nowrap>
	  <tt>-o/--output-dir &lt;output_dir_name&gt;</tt>
	  </td><td VALIGN=top>
	  The name of the directory in which you'd like Spats to write its output. Default <tt>spats_out</tt>.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--adapter-t &lt;adapter sequence&gt;</tt>
	  </td><td VALIGN=top>
	  The sequence of adapter_t (which contains the RT primer).  Example <tt>AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATC</tt>
	  </td></tr>
		
	  <tr><td VALIGN=top nowrap>
	  <tt>--adapter-b &lt;adapter sequence&gt;</tt>
	  </td><td VALIGN=top>
	  The sequence of adapter_b (which is ligated to the single-stranded cDNA products of the RT reaction).  Example <tt>AGATCGGAAGAGCGGTTCAGCAGGAATGCCGAGACCGATCTCGTATGCCGTCTTCTGCTTG</tt>
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--num-mismatches &lt;0-3&gt;</tt>
	  </td><td VALIGN=top>
	  The number of mismatches you want Bowtie to allow in its alignments for each read to the target RNAs.
	  </td></tr>
	
	  <tr><td VALIGN=top nowrap>
	  <tt>--num-threads &lt;threads&gt;</tt>
	  </td><td VALIGN=top>
	  The number of threads you want Bowtie to use when aligning the reads to the target RNAs
	  </td></tr>
	  </table><br/>
	
	  <h2 id="spats_input">Spats Input</h2><br/>
		<p>
		Spats takes three files as input: the target RNA file, and two files containing the paired-end sequencing
        reads from the experiment in FASTQ format. The sequecing reads for both the plus and minus channels of the
        SHAPE experiment must be contained in the files. Please refer to the references on the right hand side for
        more information on how these handles are included in library preparation and sequencing. It is critical
        that the mated reads appear in the same order in each file. The target RNA file is in FASTA format, and
        contains one entry for each RNA that is included in the experiment, e.g.
		</p>
		<br/>
		<pre>
		>target_GTAC
		GGTCGTGCCTAGCGAAGTCATAAGCTAGGGCAGTCTTTAGAGGCTGACGGCAGGAAAAAAG
		CCTACGTCTTCGGATATGGCTGAGTATCCTTGAAAGTGCCACAGTGACGAAGTCTCACTAG
		AAATGGTGAGAGTGGAACGCGGTAAACCCCTCGACCGATCCGCTTCGGCGGATCCGTACAA
		ATCGGGCTTCGGTCCGGTTC
		>target_TTAA
		GGTCGTGCCTAGCGAAGTCATAAGCTAGGGCAGTCTTTAGAGGCTGACGGCAGGAAAAAAG
		CCTACGTCTTCGGATATGGCTGAGTATCCTTGAAAGTGCCACAGTGACGAAGTCTCACTAG
		AAATGGTGAGAGTGGAACGCGGTAAACCCCTCGACCGATCCGCTTCGGCGGATCCTTAAAA
		ATCGGGCTTCGGTCCGGTTC
		...
		</pre>
		In this case, two identical RNA sequences were barcoded with two sequences (GTAC, TTAA) and run in the same
        experiment.
      <br/>
      <br/>
  <h2 id="spats_output">Spats Output</h2><br/>
	
	Spats produces a directory (by default called <tt>spats_out</tt>) containing several output files.
	<h3>De-multiplexed reads</h3>
	Spats splits the reads into the plus and minus channels according to the 
	handle sequences you've specified.  The output directory will contain FASTQ
	files named <tt>&lt;treated_handle&gt;_1.fq</tt>, <tt>&lt;treated_handle&gt;_2.fq</tt>, <tt>&lt;untreated_handle&gt;_1.fq</tt> and <tt>&lt;untreated_handle&gt;_2.fq</tt> 
	<h3>Reactivities</h3>
		<p>
		The <tt>reactivities.out</tt> file is a tab-delimited file of values 
		providing the reactivity at each base of the target RNAs.  The example 
		below shows the first few rows of a run against wild-type RNaseP:
		</p>
		<br/>
	<pre>
	sequence  five_prime_offset	nucleotide	treated_mods	untreated_mods	beta  	    theta	c
	RNaseP_WT	0	            *	           77211	55078	        -	    -	        0.840988
	RNaseP_WT	1	            G	           10779	5532	        0.0422978   0.041584	0.840988
	RNaseP_WT	2	            G	           5505	        3141	        0.012441    0.0120808	0.840988
	RNaseP_WT	3	            T	           289	        249	        0	    0	        0.840988
	RNaseP_WT	4	            C	           381	        306	        0	    0	        0.840988
	...
	</pre>
	<table CELLSPACING=15>
	
	<tr><td VALIGN=top nowrap>sequence</td><td VALIGN=top>	Name of target RNA (from the input sequences FASTA file.)</td></tr>
	<tr><td VALIGN=top nowrap>five_prime_offset</td><td VALIGN=top>	Position in target RNA, 0=full length</td></tr>
	<tr><td VALIGN=top nowrap>nucleotide</td><td VALIGN=top>	Base at position in target, *=full length</td></tr>
	<tr><td VALIGN=top nowrap>treated_mods</td><td VALIGN=top>	Number of (+) fragments ending at position in target</td></tr>
	<tr><td VALIGN=top nowrap>untreated_mods</td><td VALIGN=top>	Number of (-) fragments ending at position in target</td></tr>
	<tr><td VALIGN=top nowrap>beta</td><td VALIGN=top>	Beta value at position in target</td></tr>
	<tr><td VALIGN=top nowrap>theta	</td><td VALIGN=top>Theta value at position in target</td></tr>
	<tr><td VALIGN=top nowrap>c</td><td VALIGN=top>	Estimated SHAPE modifications per target</td></tr>
	</table>
	See the references for definitions of beta, theta and c.
	</div>
  </div>
  <div id="footer">
  	<table width="100%" cellspacing=15><tr><td>
		<td> This research was supported in part by
            the Miller Institute for Basic Research in Science at UC
            Berkeley. </td>
    </td><td align=right>
    Administrator: <a href="mailto:cole@cs.umd.edu">Cole Trapnell</a>. Design by <a href="http://www.free-css-templates.com" title="Design by David Herreman">David Herreman</a>
    </td></tr></table>
  </div>
</div>

<!-- Google analytics code -->
<!-- <script type="text/javascript">
var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
</script>
<script type="text/javascript">
try {
var pageTracker = _gat._getTracker("UA-6101038-2");
pageTracker._trackPageview();
} catch(err) {}</script> -->
<!-- End Google analytics code -->
</body>
</html>
